resource/graph_node.go

```diff
@@ -356,24 +397,35 @@ func (w *GraphNode) replace(other *GraphNode) error {
 	w.current = other.current
 	w.currentModel = other.currentModel
 	w.config = other.config
-	w.needsReconfigure = other.needsReconfigure
 	w.lastErr = other.lastErr
-	w.markedForRemoval = other.markedForRemoval
 	w.unresolvedDependencies = other.unresolvedDependencies
 	w.needsDependencyResolution = other.needsDependencyResolution
 
+	w.state = other.state
+	w.transitionedAt = other.transitionedAt
+
 	// other is now owned by the graph/node and is invalidated
 	other.updatedAt = 0
 	other.graphLogicalClock = nil
 	other.lastReconfigured = nil
 	other.current = nil
 	other.currentModel = Model{}
 	other.config = Config{}
-	other.needsReconfigure = false
 	other.lastErr = nil
-	other.markedForRemoval = false
 	other.unresolvedDependencies = nil
 	other.needsDependencyResolution = false
+
+	other.state = NodeStateUnconfigured
+	other.transitionedAt = time.Time{}
+
 	other.mu.Unlock()
 	return nil
 }
+
+func (w *GraphNode) transitionTo(state NodeState) {
+	if w.state == state {
+		return
+	}
+	w.state = state
+	w.transitionedAt = time.Now()
+}
```

Comment:
no concrete follow-up, but two notes/thoughts:
* we could consider expanding this helper to validate if a transition is legal
* it might be interesting to make self-transitions error to expose scenarios where need a more granular state

Reply:
agreed, I think we should eventually use this helper to validate if a transition is legal

Reply:
Really like both those thoughts. Why not add both of those now?

Reply:
if we're worried that some transitions are illegal, we should log a warning for now and try to root it out

Reply:
Logging for now SGTM.

Reply:
great idea, will add warning logs in these cases

Reply:
I didn't dive too deep into this, but here are the possibly illegal transitions that happen in our test suite. i'm not counting self-transitions since it's clear if that should be okay yet.

|Transition|Possible reason|
|--|--|
|Configuring -> Removing |Seems like resources can "abort" configuration if they are marked for removal. maybe this is fine. |
| Removing -> Configuring | Allowed due to [this logic](https://github.com/viamrobotics/rdk/blob/053632958fea4fbff817bd2ec3a420265fc50ac9/resource/graph_node.go#L259-L276), which forces nodes marked for removal to be configured instead. |
| Removing -> Ready | This one is weird - unclear why how this can happen.|